cherab/solps/formats/balance.py

```diff
@@ -0,0 +1,192 @@
+from raysect.core.math import Discrete2DMesh
+
+from cherab.core.math.mappers import AxisymmetricMapper
+from cherab.core.atomic.elements import hydrogen, deuterium, helium, beryllium, carbon, nitrogen, oxygen, neon, \
+    argon, krypton, xenon
+
+from cherab.solps.eirene import Eirene
+from cherab.solps.mesh_geometry import SOLPSMesh
+from cherab.solps.solps_plasma import SOLPSSimulation
+
+Q = 1.602E-19
+
+# key is nuclear charge Z and atomic mass AMU
+_popular_species = {
+    (1, 2): deuterium,
+    (6, 12.0): carbon,
+    (2, 4.003): helium,
+    (7, 14.0): nitrogen,
+    (10, 20.180): neon,
+    (18, 39.948): argon,
+    (18, 40.0): argon,
+    (36, 83.798): krypton,
+    (54, 131.293): xenon
+}
+
+
+# Code developed by J. Harrison 9/4/2019
+def load_solps_from_balance(balance_filename, debug=False):
+    """
+    Load a SOLPS simulation from SOLPS balance.nc output files.
+
+    """
+
+    if not os.path.isdir(balance_filename):
+        RuntimeError("file name must be valid")
+    
+    # Open the file
+    fhandle = netcdf.netcdf_file(balance_filename,'r')
+	
+    # Load SOLPS mesh geometry
+    cr_x = copy.deepcopy(fhandle.variables['crx'].data)
+    cr_z = copy.deepcopy(fhandle.variables['cry'].data)
+    vol = copy.deepcopy(fhandle.variables['vol'].data)
+	
+    # Re-arrange the array dimensions in the way CHERAB expects...
+    cr_x = swapaxes(cr_x,0,2)
+    cr_x = swapaxes(cr_x,0,1)
+    cr_z = swapaxes(cr_z,0,2)
+    cr_z = swapaxes(cr_z,0,1)
+	
+    # Create the SOLPS mesh
+    mesh = SOLPSMesh(cr_x,cr_z,vol)	
+
+    sim = SOLPSSimulation(mesh)
+    ni = mesh.nx
+    nj = mesh.ny
+
+    # TODO: add code to load SOLPS velocities and magnetic field from files
+
+    # Load electron species
+    sim._electron_temperature = copy.deepcopy(fhandle.variables['te'].data)/Q
+    sim._electron_density = copy.deepcopy(fhandle.variables['ne'].data)
+
+    ##########################################
+    # Load each plasma species in simulation #
+    ##########################################
+
+    sim._species_list = []
+    n_species = len(fhandle.variables['am'].data)
+	
+    for i in range(n_species):
+
+        # Extract the nuclear charge	
+        if fhandle.variables['species'].data[i,1] == b'D':
+	        zn = 1
+        if fhandle.variables['species'].data[i,1] == b'C':
+            zn = 6
+        if fhandle.variables['species'].data[i,1] == b'N':
+            zn = 7
+        if fhandle.variables['species'].data[i,1] == b'N' and fhandle.variables['species'].data[i,2] == b'e':
+            zn = 10
+        if fhandle.variables['species'].data[i,1] == b'A' and fhandle.variables['species'].data[i,2] == b'r':
+            zn = 18
+	
+        am = float(fhandle.variables['am'].data[i])  # Atomic mass
+        charge = int(fhandle.variables['za'].data[i])  # Ionisation/charge
+        species = _popular_species[(zn, am)]
+		
+        # If we only need to populate species_list, there is probably a faster way to do this...		
```

Comment:
Since the number of species is typically small, you probably don't need to worry about performance. Your intent here is clear: that's more important.